pkg/storage/stores/shipper/indexgateway/ringmanager.go

```diff
@@ -0,0 +1,249 @@
+package indexgateway
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	loki_util "github.com/grafana/loki/pkg/util"
+)
+
+const (
+	ringAutoForgetUnhealthyPeriods = 10
+	ringNameForServer              = "index-gateway"
+	ringNumTokens                  = 128
+	ringCheckPeriod                = 3 * time.Second
+
+	// RingIdentifier is used as a unique name to register the Index Gateway ring.
+	RingIdentifier = "index-gateway"
+
+	// RingKey is the name of the key used to register the different Index Gateway instances in the key-value store.
+	RingKey = "index-gateway"
+)
+
+type ManagerMode int
+
+const (
+	ClientMode ManagerMode = iota
+	ServerMode
+)
+
+type ringTenantBuffers struct {
+	bufDescs []ring.InstanceDesc
+	bufHosts []string
+	bufZones []string
+}
+
+type RingManager struct {
+	services.Service
+
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	RingLifecycler *ring.BasicLifecycler
+	Ring           *ring.Ring
+	managerMode    ManagerMode
+
+	buffers ringTenantBuffers
+	cfg     Config
+
+	log log.Logger
+}
+
+func NewRingManager(managerMode ManagerMode, cfg Config, log log.Logger, registerer prometheus.Registerer) (*RingManager, error) {
+	rm := &RingManager{
+		cfg: cfg, log: log, managerMode: managerMode,
+	}
+
+	if cfg.Mode != RingMode {
+		return rm, nil
```

Comment:
I think we should return an error because the code might panic or some other problems can happen which might make it hard to understand the real problem.